File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    vector<string> res;
    int size;
    void helper(string s, int o, int c) {
        if(s.size() == 2 * size) {
            res.push_back(s);
            return;
        }
        if(c >= o && o != 0) {
            helper(s + '(', o - 1, c);
        }
        if(o < c && c != 0) {
            helper(s + ')', o, c - 1);
        }
    }
    vector<string> generateParenthesis(int n) {
        size = n;
        helper("", n, n);
        return res;
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result = []
        
        def backtrack(current, open_count, close_count):
            if open_count == n and close_count == n:
                result.append(current)
                return
            
            if open_count < n:
                backtrack(current + '(', open_count + 1, close_count)
            if close_count < open_count:
                backtrack(current + ')', open_count, close_count + 1)
        
        backtrack("", 0, 0)
        return result